docker/docker_client.go

```diff
@@ -195,7 +195,9 @@ func dockerCertDir(sys *types.SystemContext, hostPort string) (string, error) {
 
 // newDockerClientFromRef returns a new dockerClient instance for refHostname (a host a specified in the Docker image reference, not canonicalized to dockerRegistry)
 // “write” specifies whether the client will be used for "write" access (in particular passed to lookaside.go:toplevelFromSection)
-func newDockerClientFromRef(sys *types.SystemContext, ref dockerReference, write bool, actions string) (*dockerClient, error) {
+func newDockerClientFromRef(ctx context.Context, sys *types.SystemContext,
+	ref dockerReference, write bool, actions string) (*dockerClient, error) {
+
 	registry := reference.Domain(ref.ref)
 	username, password, err := config.GetAuthentication(sys, reference.Domain(ref.ref))
```

Comment:
Okay, we could introduce an additional config option for the mirror to specify if the mirror needs the same authentication credentials as the upstream registry.

Reply:
I'm not sure how to approach this, one option could be to add further configuration options for mirrors. Then it would be possible to distinguish the auth settings for the registries/mirrors.

Reply:
This sounds like a case for [Proxy-Authenticate][1] (if the mirror is already authorized to access the upstream-registry content)?

Or does the mirror have no upstream-registry authorization on its own and want the user to delegate their upstream-registry access?  In that case, it sounds like we want challenge/response approach so users could delegate to the mirror on a per-request basis instead of just handing over their secret.  But in order for that to work, the upstream registry would need to support challenge-response authentication, and we'd need to revisit the Docker-credential API (which is currently URI-authority-keyed with no way to forward challenges).

[1]: https://tools.ietf.org/html/rfc7235#section-4.3

Reply:
The way I think about a mirror (which may be not how things work in practice), a mirror is something very different from a proxy (transparent or not).

With a proxy, the client needs to supply credentials for the upstream server, and the proxy may or may not have a local copy of the data — but it is _always_ expected to return data ~equivalent to what the upstream server would return.

With a mirror, the _mirror administrator_ is in control of what the mirror is serving; while, sure, that may be some kind of proxy-like on-line sync, it may be a periodic process (only update once a week to save on bandwidth), it may be some manual gating (only get a new version of the base image after it has been tested), or it may be a manual data transfer to an off-line data center.  In most of these situations it does not make all that much sense to send credentials to the upstream registry to the mirror, because the mirror has typically already fetched the image using its own credentials as configured by the admin; on the contrary, the mirror may have its own authentication system with quite different policies (e.g. only users of a single department are allowed to use a pay-per-use product, but those users are not given credentials to download directly from the ISV).

A mirror wouldn’t typically a very ordinary docker/distribution server — only combined with a process to, at $some time, `skopeo copy` images from the upstream to the mirror. Such a server can, AFAICS, perfectly fine use the default docker/distribution token mechanism without any modification; the only concern is how to express this / look this up in the client’s configuration.

Reply:
> A mirror wouldn’t typically a very ordinary docker/distribution server

“A mirror would typically be …”